doc: improve readme (#1732)
intel/torch-xpu-ops: README.md

# Torch XPU Operators*

> Torch XPU Operators* project is integral part of [PyTorch](https://github.com/pytorch/pytorch) to support XPU acceleration backend. PyTorch build system automatically clones this repository at the pin pointed commit, branch or tag specified in the following file:

Reviewer: ```suggestion
Torch XPU Operators* project is integral an part of [PyTorch](https://github.com/pytorch/pytorch) to support XPU acceleration backend. The PyTorch build system automatically clones this repository at the commit, branch, or tag pinned and specified in the following file of the [PyTorch repository](https://github.com/pytorch/pytorch):
```

* https://github.com/pytorch/pytorch/blob/main/third_party/xpu.txt

Cloned copy becomes available at `./third_party/torch-xpu-ops/` relative to the root of the checked out PyTorch tree.

Torch XPU Operators* implements some of the operators for Intel GPU devices accessible via PyTorch XPU acceleration backend:

* PyTorch ATen operators
* Torchvision operators

> Most operators are implemented as SYCL kernels sources of which are available within this repository. Some operators (linear algebra)
> are implemented thru the calls to [oneMKL](https://www.intel.com/content/www/us/en/developer/tools/oneapi/onemkl.html#gs.mdl4z3) library.

Reviewer: ```suggestion
Most operators are implemented as SYCL kernels, the sources of which are available in this repository. Some operators (linear algebra)
are implemented through calls to the [Intel® oneAPI Math Kernel Library (oneMKL)](https://www.intel.com/content/www/us/en/developer/tools/oneapi/onemkl.html).
```

> Note that few operators (convolution and gemm) for XPU backend are implemented directly in PyTorch sources thru the calls to [oneDNN](https://github.com/uxlfoundation/oneDNN) library. These sources can be found at https://github.com/pytorch/pytorch/tree/main/aten/src/ATen/native/mkldnn/xpu folder.

Reviewer: ```suggestion
Note that a few operators (convolution and matrix-matrix multiplication (`gemm`)) for the PyTorch XPU backend are implemented directly in PyTorch sources through calls to the [oneAPI Deep Neural Network Library (oneDNN)](https://github.com/uxlfoundation/oneDNN). These sources can be found in the PyTorch repository at https://github.com/pytorch/pytorch/tree/main/aten/src/ATen/native/mkldnn/xpu.
```

* Intel GPU Driver: Install Intel GPU drivers along with compute and media runtimes and development packages.
* Intel® Deep Learning Essentials: Install a subset of Intel® oneAPI components needed for building and running PyTorch.

## Build and install

> This project can't be built or installed stand alone. Project is used when PyTorch is built with XPU backend support.

Reviewer: ```suggestion
This project cannot be built or installed as a stand-alone. This project gets built when PyTorch is built with XPU backend support.
```

**To install PyTorch with XPU backend from pre-built binary packages**, use one of the available distribution channels:

* For release builds:

```
pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/xpu
```

* For nightly builds

```
pip3 install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/xpu
```

> **To build PyTorch with XPU backend from sources**, refer to PyTorch [documentation](https://github.com/pytorch/pytorch/blob/main/README.md#from-source) and pay attention on [Intel GPU Support](https://github.com/pytorch/pytorch/blob/main/README.md#intel-gpu-support) section. Shortly, PyTorch build can be triggered as follows:

Reviewer: ```suggestion
**To build PyTorch with XPU backend from sources**, refer to the [Intel GPU Support](https://github.com/pytorch/pytorch/blob/main/README.md#intel-gpu-support) section of the [PyTorch documentation](https://github.com/pytorch/pytorch/blob/main/README.md#from-source). In summary, the PyTorch build for XPU backend can be triggered as follows:
```

```bash
git clone https://github.com/pytorch/pytorch.git && cd pytorch
pip install -r requirements.txt
python setup.py install >log.txt 2>&1
```

> Monitor for the following log printout indicating that XPU support is being built:

Reviewer: ```suggestion
Look for the following lines in the log file indicating that XPU backend is being built:
```

```
$ cat log.txt | grep -E "(USE_XPU|USE_XCCL)\s*:"
--   USE_XPU               : 1

> --   USE_XCCL              : ON

Reviewer: Thanks for the clarification. We can keep it as is then.

```

> If building from sources, note the following environment variables which control PyTorch XPU backend build:

Reviewer: ```suggestion
If building from sources, note the following environment variables which control the PyTorch XPU backend build:
```

| Environment variable | Default | Notes |
| --- | --- | --- |
| `USE_XPU` | `ON` | Enables XPU backend support |
| `USE_XCCL` | `ON` (>= PT2.8) | Enables XCCL distributed backend support |
| `TORCH_XPU_ARCH_LIST` | depends on the PT and OS versions | Build SYCL kernels for specified platform(s) |

> The `TORCH_XPU_ARCH_LIST` allows to specify comma separated list of platforms for which SYCL kernels will be built. This helps to reduce build time. Option does not affect oneDNN or oneMKL based operators. Note that if PyTorch will be executed on the platform for which SYCL kernels were not built, they will be JIT compiled.

Reviewer: ```suggestion
The `TORCH_XPU_ARCH_LIST` allows to specify a comma separated list of platforms for which SYCL kernels will be built. This helps to reduce the build time. This option does not affect oneDNN- or oneMKL-based operators. Note that if PyTorch is executed on a platform that the SYCL kernels were not built for, then they are compiled just-in-time (JIT).
```

> ## Verify

Reviewer: ```suggestion
## Verification
```

Once PyTorch is built or installed, verify that PyTorch XPU backend is available as follows:

```
$ python -c "import torch; print(torch.xpu.is_available())"
True
python3 -c "import torch; print(torch.distributed.distributed_c10d.is_xccl_available())"
True
```

# FAQ

Some confusion might arise about relationship of this repostory and PyTorch XPU backend in general with [Intel Extension for PyTorch (IPEX)](https://github.com/intel/intel-extension-for-pytorch). See answers on that below.

**Does PyTorch XPU backend implementation uses IPEX?**

No. PyTorch XPU backend implementation does not use IPEX and can be used without IPEX levaraging:

- Standard PyTorch API to access eager mode operators, compile modules (including Triton kernels), profiling, etc.
- SYCL kernels to enhance applications with custom operators via PyTorch CPP Extension API

**Does IPEX uses PyTorch XPU backend implementation?**

> Yes. IPEX relies on PyTorch XPU backend implementation and augments it with additional features and operators. On top of that IPEX implements select features and operators outside of standard PyTorch API which are required in popular AI frameworks such as vLLM, Huggingface TGI, SGLang, and others. The ultimate long term goal is to upstream or substituite with better upstream implementations as many IPEX code as possible. Each next IPEX release is a step toward that goal as less and less features are being implemented in IPEX and more are taken from upstream projects. Some of these changes (for example, implemetation of PyTorch eager mode operators) are made transparent for the end users. Other changes (for example, a switch to `xccl` distributed backend) are associated with the API changes toward usaging APIs defined in the upstream projects.

Author: I will drop last 2 sentences as they indeed might be confusing and will require to add too much text to thoroughly explain.

> I also did not understand the last sentence of this answer. Could you clarify?

There are IPEX specific APIs which are getting replaced by PyTorch APIs. Distributed backend is one of such examples. It changes a name - in IPEX its analogues was called "ccl", in pytorch it's "xccl". And there are other API level differences in its usage.

> Yes. IPEX relies on PyTorch XPU backend implementation and augments it with additional features and operators. On top of that IPEX implements select features and operators outside of standard PyTorch API which are required in popular AI frameworks such as vLLM, Huggingface TGI, SGLang, and others. The ultimate long term goal is to upstream or substituite with better upstream implementations as many IPEX code as possible. Each next IPEX release is a step toward that goal as less and less features are being implemented in IPEX and more are taken from upstream projects. Some of these changes (for example, implemetation of PyTorch eager mode operators) are made transparent for the end users. Other changes (for example, a switch to `xccl` distributed backend) are associated with the API changes toward usaging APIs defined in the upstream projects.

Reviewer: I see, I understand better now. I'm okay with dropping the last two sentences. Feel free to resolve this thread (I don't have the option to resolve).

```
$ cat log.txt | grep -E "(USE_XPU|USE_XCCL)\s*:"
--   USE_XPU               : 1
--   USE_XCCL              : ON
```

If building from sources, note the following environment variables which control the PyTorch XPU backend build:

| Environment variable | Default | Notes |
| --- | --- | --- |
| `USE_XPU` | `ON` | Enables XPU backend support |
| `USE_XCCL` | `ON` (>= PT2.8) | Enables XCCL distributed backend support |
| `TORCH_XPU_ARCH_LIST` | depends on the PT and OS versions | Build SYCL kernels for specified platform(s) |

The `TORCH_XPU_ARCH_LIST` allows to specify a comma separated list of platforms for which SYCL kernels will be built. This helps to reduce the build time. This option does not affect oneDNN- or oneMKL-based operators. Note that if PyTorch is executed on a platform that the SYCL kernels were not built for, then they are compiled just-in-time (JIT).

## Verification

Once PyTorch is built or installed, verify that PyTorch XPU backend is available as follows:

```
$ python3 -c "import torch; print(torch.xpu.is_available())"
True
python3 -c "import torch; print(torch.distributed.distributed_c10d.is_xccl_available())"
True
```

# FAQ

> Some confusion might arise about relationship of this repostory and PyTorch XPU backend in general with [Intel® Extension for PyTorch (IPEX)](https://github.com/intel/intel-extension-for-pytorch). See answers on that below.

Reviewer 2: IPEX is not an valid name, please use full name, not IPEX. And Intel Extension for PyTorch is deprecating, even not highlight the question is fine here.

```
$ cat log.txt | grep -E "(USE_XPU|USE_XCCL)\s*:"
--   USE_XPU               : 1
--   USE_XCCL              : ON
```

If building from sources, note the following environment variables which control the PyTorch XPU backend build:

| Environment variable | Default | Notes |
| --- | --- | --- |
| `USE_XPU` | `ON` | Enables XPU backend support |
| `USE_XCCL` | `ON` (>= PT2.8) | Enables XCCL distributed backend support |
| `TORCH_XPU_ARCH_LIST` | depends on the PT and OS versions | Build SYCL kernels for specified platform(s) |

The `TORCH_XPU_ARCH_LIST` allows to specify a comma separated list of platforms for which SYCL kernels will be built. This helps to reduce the build time. This option does not affect oneDNN- or oneMKL-based operators. Note that if PyTorch is executed on a platform that the SYCL kernels were not built for, then they are compiled just-in-time (JIT).

## Verification

Once PyTorch is built or installed, verify that PyTorch XPU backend is available as follows:

```
$ python3 -c "import torch; print(torch.xpu.is_available())"
True
python3 -c "import torch; print(torch.distributed.distributed_c10d.is_xccl_available())"
True
```

> # FAQ

Reviewer 3: From my understanding, the FAQ should reside within the IPEX repository, and there's no need to explicitly link it to IPEX elsewhere. 

By the way, I'm not entirely clear on IPEX's current positioning—could you please confirm this with the IPEX owner?
